From 380a3e73ad5a383bc491302f99513746ba4bd6a4 Mon Sep 17 00:00:00 2001
From: Patrice Chotard <patrice.chotard@foss.st.com>
Date: Fri, 20 Aug 2021 10:53:23 +0200
Subject: [PATCH 051/120] misc: Add STM32 Octo Memory Manager (OMM) support

This driver manages the muxing between the 2 OSPI busses and
the 2 output ports. There are 4 possible muxing configurations:
  - direct mode (no multiplexing): OSPI1 output is on port 1 and OSPI2
    output is on port 2
  - OSPI1 and OSPI2 are multiplexed over the same output port 1
  - swapped mode (no multiplexing), OSPI1 output is on port 2,
    OSPI2 output is on port 1
  - OSPI1 and OSPI2 are multiplexed over the same output port 2

It also manages :
  - the split of the memory area shared between the 2 OSPI instances.
  - chip select selection override.
  - the time between 2 transactions in multiplexed mode.
  - OMI resource access check and probe its child depending of these
    access right.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git e8bd828eab0383929fa026ec1b5af05166756f07 misc: Add STM32 Octo Memory Manager (OMM) support]
Signed-off-by: Patrice Chotard <patrice.chotard@foss.st.com>
Change-Id: I15307bead4556dd3af823579e8145bd450404807
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/230295
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Christophe KERELLO <christophe.kerello@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/308511
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Christophe KERELLO <christophe.kerello@st.com>
---
 drivers/misc/Kconfig     |  17 ++
 drivers/misc/Makefile    |   1 +
 drivers/misc/stm32_omm.c | 389 +++++++++++++++++++++++++++++++++++++++
 3 files changed, 407 insertions(+)
 create mode 100644 drivers/misc/stm32_omm.c

diff --git a/drivers/misc/Kconfig b/drivers/misc/Kconfig
index 8b8f6309ada..9aaa5833e06 100644
--- a/drivers/misc/Kconfig
+++ b/drivers/misc/Kconfig
@@ -487,6 +487,23 @@ config K3_FUSE
 	  If you say Y here, you will get support for the fuse API (OTP)
 	  for TI K3 architecture.
 
+config STM32_OMM
+	bool "Enable Octo Memory Manager (OMM) driver for the STM32MP2 SoC's family"
+	depends on STM32MP25X && MISC
+	help
+	  This driver manages the muxing between the 2 OSPI busses and
+	  the 2 output ports. There are 4 possible muxing configurations:
+	  - direct mode (no multiplexing): OSPI1 output is on port 1 and OSPI2
+	       output is on port 2
+	  - OSPI1 and OSPI2 are multiplexed over the same output port 1
+	  - swapped mode (no multiplexing), OSPI1 output is on port 2,
+	       OSPI2 output is on port 1
+	  - OSPI1 and OSPI2 are multiplexed over the same output port 2
+	  It also manages :
+	    - the split of the memory area shared between the 2 OSPI instances.
+	    - chip select selection override.
+	    - the time between 2 transactions in multiplexed mode.
+
 config STM32_RCC
 	bool "Enable RCC driver for the STM32 SoC's family"
 	depends on (ARCH_STM32 || ARCH_STM32MP) && MISC
diff --git a/drivers/misc/Makefile b/drivers/misc/Makefile
index 248068d5b43..a5bfc44e8d0 100644
--- a/drivers/misc/Makefile
+++ b/drivers/misc/Makefile
@@ -77,6 +77,7 @@ obj-$(CONFIG_SIFIVE_OTP) += sifive-otp.o
 obj-$(CONFIG_SMSC_LPC47M) += smsc_lpc47m.o
 obj-$(CONFIG_SMSC_SIO1007) += smsc_sio1007.o
 obj-$(CONFIG_STM32MP_FUSE) += stm32mp_fuse.o
+obj-$(CONFIG_STM32_OMM) += stm32_omm.o
 obj-$(CONFIG_STM32_RCC) += stm32_rcc.o
 obj-$(CONFIG_SYS_DPAA_QBMAN) += fsl_portals.o
 obj-$(CONFIG_TEGRA186_BPMP) += tegra186_bpmp.o
diff --git a/drivers/misc/stm32_omm.c b/drivers/misc/stm32_omm.c
new file mode 100644
index 00000000000..cdd3cef1d3c
--- /dev/null
+++ b/drivers/misc/stm32_omm.c
@@ -0,0 +1,389 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) 2021, STMicroelectronics - All Rights Reserved
+ */
+
+#include <clk.h>
+#include <dm.h>
+#include <regmap.h>
+#include <stm32_omi.h>
+#include <syscon.h>
+#include <dm/device_compat.h>
+#include <dm/device-internal.h>
+#include <dm/lists.h>
+#include <dm/of_addr.h>
+#include <dm/of_access.h>
+#include <linux/bitfield.h>
+#include <linux/ioport.h>
+#include <mach/rif.h>
+
+#define OCTOSPIM_CR		0
+#define CR_MUXEN		BIT(0)
+#define CR_MUXENMODE_MASK	GENMASK(1, 0)
+#define CR_CSSEL_OVR_EN		BIT(4)
+#define CR_CSSEL_OVR_MASK	GENMASK(6, 5)
+#define CR_REQ2ACK_MASK		GENMASK(23, 16)
+
+#define OMM_CHILD_NB		2
+
+#define NSEC_PER_SEC 1000000000L
+
+struct stm32_omm_plat {
+	struct regmap *omm_regmap;
+	struct regmap *syscfg_regmap;
+	struct clk clk;
+	struct reset_ctl reset_ctl;
+	resource_size_t mm_ospi2_size;
+	u32 mux;
+	u32 cssel_ovr;
+	u32 req2ack;
+	u32 amcr_base;
+	u32 amcr_mask;
+	unsigned long clk_rate_max;
+};
+
+static int stm32_omm_set_amcr(struct udevice *dev, bool set)
+{
+	struct stm32_omm_plat *plat = dev_get_plat(dev);
+	unsigned int amcr, read_amcr;
+
+	amcr = plat->mm_ospi2_size / SZ_64M;
+
+	if (set)
+		regmap_update_bits(plat->syscfg_regmap, plat->amcr_base,
+				   plat->amcr_mask, amcr);
+
+	/* read AMCR and check coherency with memory-map areas defined in DT */
+	regmap_read(plat->syscfg_regmap, plat->amcr_base, &read_amcr);
+	read_amcr = read_amcr >> (ffs(plat->amcr_mask) - 1);
+
+	return amcr != read_amcr;
+}
+
+static int stm32_omm_configure(struct udevice *dev)
+{
+	struct stm32_omm_plat *plat = dev_get_plat(dev);
+	int ret;
+	u32 mux = 0;
+	u32 cssel_ovr = 0;
+	u32 req2ack = 0;
+
+	ret = clk_enable(&plat->clk);
+	if (ret) {
+		dev_err(dev, "Failed to enable OMM clock (%d)\n", ret);
+		return ret;
+	}
+
+	reset_assert(&plat->reset_ctl);
+	udelay(2);
+	reset_deassert(&plat->reset_ctl);
+
+	if (plat->mux & CR_MUXEN) {
+		if (!plat->req2ack) {
+			req2ack = DIV_ROUND_UP(plat->req2ack,
+					       NSEC_PER_SEC / plat->clk_rate_max) - 1;
+			if (req2ack > 256)
+				req2ack = 256;
+		}
+
+		req2ack = FIELD_PREP(CR_REQ2ACK_MASK, req2ack);
+		regmap_update_bits(plat->omm_regmap, OCTOSPIM_CR,
+					 CR_REQ2ACK_MASK, req2ack);
+	}
+
+	if (plat->cssel_ovr != 0xff) {
+		cssel_ovr = FIELD_PREP(CR_CSSEL_OVR_MASK, cssel_ovr);
+		cssel_ovr |= CR_CSSEL_OVR_EN;
+		regmap_update_bits(plat->omm_regmap, OCTOSPIM_CR,
+				   CR_CSSEL_OVR_MASK, cssel_ovr);
+	}
+
+	mux = FIELD_PREP(CR_MUXENMODE_MASK, plat->mux);
+	regmap_update_bits(plat->omm_regmap, OCTOSPIM_CR,
+			   CR_MUXENMODE_MASK, mux);
+
+	return stm32_omm_set_amcr(dev, true);
+}
+
+static int stm32_omm_disable_child(struct udevice *dev, ofnode child)
+{
+	struct regmap *omi_regmap;
+	struct clk omi_clk;
+	int ret;
+
+	ret = regmap_init_mem(child, &omi_regmap);
+	if (ret) {
+		dev_err(dev, "Regmap failed for node %s\n", ofnode_get_name(child));
+		return ret;
+	}
+
+	/* retrieve OMI clk */
+	ret = clk_get_by_index_nodev(child, 0, &omi_clk);
+	if (ret) {
+		dev_err(dev, "Failed to get clock for %s\n", ofnode_get_name(child));
+		return ret;
+	}
+
+	ret = clk_enable(&omi_clk);
+	if (ret) {
+		dev_err(dev, "Failed to enable clock for %s\n", ofnode_get_name(child));
+		goto clk_free;
+	}
+
+	regmap_update_bits(omi_regmap, OSPI_CR, OSPI_CR_EN, 0);
+
+	clk_disable(&omi_clk);
+clk_free:
+
+	return ret;
+}
+
+static int stm32_omm_probe(struct udevice *dev) {
+	ofnode child_list[OMM_CHILD_NB];
+	ofnode child;
+	int ret;
+	u8 nb_child = 0;
+	u8 child_access_granted = 0;
+	u8 i;
+	bool child_access[OMM_CHILD_NB];
+
+	/* check child's access */
+	for (child = ofnode_first_subnode(dev_ofnode(dev));
+	     ofnode_valid(child);
+	     child = ofnode_next_subnode(child)) {
+
+		if (nb_child > OMM_CHILD_NB) {
+			dev_err(dev, "Bad DT, found too much children\n");
+			return -E2BIG;
+		}
+
+		if (!ofnode_device_is_compatible(child, "st,stm32mp25-omi"))
+			continue;
+
+		ret = stm32_rifsc_check_access(child);
+		if (ret < 0 && ret != -EACCES)
+			return ret;
+
+		child_access[nb_child] = false;
+		if (!ret) {
+			child_access_granted++;
+			child_access[nb_child] = true;
+		}
+
+		child_list[nb_child] = child;
+		nb_child++;
+	}
+
+	/* check if OMM's ressource access is granted */
+	ret = stm32_rifsc_check_access(dev_ofnode(dev));
+	if (ret < 0 && ret != -EACCES)
+		return ret;
+
+	if (!ret) {
+		/* All child's access are granted ? */
+		if (child_access_granted == nb_child) {
+			/* Ensure both OSPI instance are disabled before configuring OMM */
+			for (i = 0; i < nb_child; i++) {
+				ret = stm32_omm_disable_child(dev, child_list[i]);
+				if (ret)
+					return ret;
+			}
+
+			ret = stm32_omm_configure(dev);
+			if (ret)
+				return ret;
+		} else {
+			dev_dbg(dev, "Can't disable Octo Memory Manager's child\n");
+		}
+	} else {
+		dev_dbg(dev, "Octo Memory Manager resource's access not granted\n");
+		/*
+		 * AMCR can't be set, so check if current value is coherent
+		 * with memory-map areas defined in DT
+		 */
+		ret = stm32_omm_set_amcr(dev, false);
+		if (ret > 0) {
+			dev_err(dev, "AMCR value not coherent with DT memory-map areas\n");
+
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static int stm32_omm_of_to_plat(struct udevice *dev)
+{
+	struct stm32_omm_plat *plat = dev_get_plat(dev);
+	static const char *mm_name[] = { "mm_ospi1", "mm_ospi2" };
+	struct resource res, res1, mm_res;
+	struct ofnode_phandle_args args;
+	struct udevice *child;
+	unsigned long clk_rate;
+	struct clk child_clk;
+	u32 mm_size = 0;
+	int ret, idx;
+	u8 i;
+
+	ret = regmap_init_mem(dev_ofnode(dev), &plat->omm_regmap);
+	if (ret) {
+		dev_err(dev, "I/O manager regmap failed\n");
+		return ret;
+	}
+
+	ret = dev_read_resource_byname(dev, "omm_mm", &mm_res);
+	if (ret) {
+		dev_err(dev, "can't get omm_mm mmap resource(ret = %d)!\n", ret);
+		return ret;
+	}
+
+	ret = reset_get_by_index(dev, 0, &plat->reset_ctl);
+	if (ret)
+		return ret;
+
+	ret = clk_get_by_index(dev, 0, &plat->clk);
+	if (ret < 0) {
+		dev_err(dev, "Can't find I/O manager clock\n");
+		return ret;
+	}
+
+	/* parse children's clock */
+	plat->clk_rate_max = 0;
+	device_foreach_child(child, dev) {
+		ret = clk_get_by_index(child, 0, &child_clk);
+		if (ret) {
+			dev_err(dev, "Failed to get clock for %s\n",
+				dev_read_name(child));
+			return ret;
+		}
+
+		clk_rate = clk_get_rate(&child_clk);
+		if (!clk_rate) {
+			dev_err(dev, "Invalid clock rate\n");
+			return -EINVAL;
+		}
+
+		if (clk_rate > plat->clk_rate_max)
+			plat->clk_rate_max = clk_rate;
+	}
+
+	plat->mux = dev_read_u32_default(dev, "st,omm-mux", 0);
+	plat->req2ack = dev_read_u32_default(dev, "st,omm-req2ack-ns", 0);
+	plat->cssel_ovr = dev_read_u32_default(dev, "st,omm-cssel-ovr", 0xff);
+	plat->mm_ospi2_size = 0;
+
+	for (i = 0; i < 2; i++) {
+		idx = dev_read_stringlist_search(dev, "memory-region-names",
+						 mm_name[i]);
+		if (idx < 0)
+			continue;
+
+		/* res1 only used on second loop iteration */
+		res1.start = res.start;
+		res1.end = res.end;
+
+		dev_read_phandle_with_args(dev, "memory-region", NULL, 0, idx,
+					   &args);
+		ret = ofnode_read_resource(args.node, 0, &res);
+		if (ret) {
+			dev_err(dev, "unable to resolve memory region\n");
+			goto clk_free;
+		}
+
+		/* check that memory region fits inside OMM memory map area */
+		if (!resource_contains(&mm_res, &res)) {
+			dev_err(dev, "%s doesn't fit inside OMM memory map area\n",
+				mm_name[i]);
+			dev_err(dev, "[0x%llx-0x%llx] doesn't fit inside [0x%llx-0x%llx]\n",
+				res.start, res.end,
+				mm_res.start, mm_res.end);
+
+			return -EFAULT;
+		}
+
+		if (i == 1) {
+			plat->mm_ospi2_size = resource_size(&res);
+
+			/* check that OMM memory region 1 doesn't overlap memory region 2 */
+			if (resource_overlaps(&res, &res1)) {
+				dev_err(dev, "OMM memory-region %s overlaps memory region %s\n",
+					mm_name[0], mm_name[1]);
+				dev_err(dev, "[0x%llx-0x%llx] overlaps [0x%llx-0x%llx]\n",
+					res1.start, res1.end, res.start, res.end);
+
+				return -EFAULT;
+			}
+		}
+
+		mm_size += resource_size(&res);
+	}
+
+	plat->syscfg_regmap = syscon_regmap_lookup_by_phandle(dev, "st,syscfg-amcr");
+	if (IS_ERR(plat->syscfg_regmap)) {
+		dev_err(dev, "Failed to get st,syscfg-amcr property\n");
+		ret = PTR_ERR(plat->syscfg_regmap);
+		goto clk_free;
+	}
+
+	ret = dev_read_u32_index(dev, "st,syscfg-amcr", 1, &plat->amcr_base);
+	if (ret) {
+		dev_err(dev, "Failed to get st,syscfg-amcr base\n");
+		goto clk_free;
+	}
+
+	ret = dev_read_u32_index(dev, "st,syscfg-amcr", 2, &plat->amcr_mask);
+	if (ret) {
+		dev_err(dev, "Failed to get st,syscfg-amcr mask\n");
+		goto clk_free;
+	}
+
+	return 0;
+
+clk_free:
+
+	return ret;
+};
+
+static int stm32_omm_bind(struct udevice *dev)
+{
+	int ret = 0, err = 0;
+	ofnode node;
+
+	for (node = ofnode_first_subnode(dev_ofnode(dev));
+	     ofnode_valid(node);
+	     node = ofnode_next_subnode(node)) {
+		const char *node_name = ofnode_get_name(node);
+
+		if (!ofnode_is_enabled(node) || stm32_rifsc_check_access(node)) {
+			dev_dbg(dev, "%s failed to bind\n", node_name);
+			continue;
+		}
+
+		err = lists_bind_fdt(dev, node, NULL, NULL,
+				     gd->flags & GD_FLG_RELOC ? false : true);
+		if (err && !ret) {
+			ret = err;
+			dev_dbg(dev, "%s: ret=%d\n", node_name, ret);
+		}
+	}
+
+	if (ret)
+		dev_dbg(dev, "Some drivers failed to bind\n");
+
+	return ret;
+}
+
+static const struct udevice_id stm32_omm_ids[] = {
+	{ .compatible = "st,stm32mp25-omm", },
+	{},
+};
+
+U_BOOT_DRIVER(stm32_omm) = {
+	.name		= "stm32_omm",
+	.id		= UCLASS_NOP,
+	.probe		= stm32_omm_probe,
+	.of_match	= stm32_omm_ids,
+	.of_to_plat	= stm32_omm_of_to_plat,
+	.plat_auto	= sizeof(struct stm32_omm_plat),
+	.bind		= stm32_omm_bind,
+};
-- 
2.47.2

